<?xml version='1.0'?>
<robot xmlns:xacro="http://wiki.ros.org/xacro">
    <xacro:macro name="ros2_control_system">
        <ros2_control name="imu" type="sensor">
            <hardware>
                <plugin>rosbot_hardware_interfaces/RosbotImuSensor</plugin>
                <param name="connection_timeout_ms">120000</param>
                <param name="connection_check_period_ms">500</param>
            </hardware>
            <sensor name="imu">
                <state_interface name="orientation.x" />
                <state_interface name="orientation.y" />
                <state_interface name="orientation.z" />
                <state_interface name="orientation.w" />
                <state_interface name="angular_velocity.x" />
                <state_interface name="angular_velocity.y" />
                <state_interface name="angular_velocity.z" />
                <state_interface name="linear_acceleration.x" />
                <state_interface name="linear_acceleration.y" />
                <state_interface name="linear_acceleration.z" />
            </sensor>
        </ros2_control>

        <ros2_control name="wheels" type="system">
            <hardware>
                <plugin>rosbot_hardware_interfaces/RosbotSystem</plugin>
                <param name="connection_timeout_ms">120000</param>
                <param name="connection_check_period_ms">500</param>

                <!-- order of velocity commands to be published in motors_cmd Float32MultiArray msg -->
                <param name="velocity_command_joint_order">
                    rear_right_wheel_joint,
                    rear_left_wheel_joint,
                    front_right_wheel_joint,
                    front_left_wheel_joint
                </param>
            </hardware>

            <joint name="front_left_wheel_joint">
                <command_interface name="velocity" />
                <state_interface name="position" />
                <state_interface name="velocity" />
            </joint>
            <joint name="front_right_wheel_joint">
                <command_interface name="velocity" />
                <state_interface name="position" />
                <state_interface name="velocity" />
            </joint>
            <joint name="rear_left_wheel_joint">
                <command_interface name="velocity" />
                <state_interface name="position" />
                <state_interface name="velocity" />
            </joint>
            <joint name="rear_right_wheel_joint">
                <command_interface name="velocity" />
                <state_interface name="position" />
                <state_interface name="velocity" />
            </joint>
        </ros2_control>
    </xacro:macro>
</robot>